ruby/phase-1.md

```diff
@@ -0,0 +1,434 @@
+# Ruby SDK - Phase 1
+
+Ruby is a popular dynamic programming language that many companies are leaning on to beat their
+competitors to the market. With the focus on the speed of execution it is very common for these
+companies to rely on managed solutions. This presents an opportunity for Temporal to become Ruby's
+de facto standard for workflow orchestration.
+
+This proposal aims to adhere to the Ruby way of writing code favouring convention over
+configuration, readability, testability and developer's productivity. These are prioritised in order
+to avoid the destiny of the [AWS Flow Framework for Ruby](https://github.com/amazon-archives/aws-flow-ruby)
+that never appealed to the Ruby community.
+
+
+## Disclaimer
+
+This proposal draws substantial inspiration from the [Temporal Ruby SDK](https://github.com/coinbase/temporal-ruby)
+developed at Coinbase by the author of this proposal. This allows us to pick the best parts of the
+former SDK and improve on the parts that are not so good.
+
+The proposal itself follows the [Python SDK proposal](https://github.com/temporalio/proposals/blob/master/python/phase-1.md)
+for the same exact reason.
+
+
+## Timeline
+
+The Ruby SDK will be implemented in four loosely defined phases:
+
+- Phase 1 — initial setup, execution attributes, workflow/activity outer APIs, client and worker
+- Phase 2 — workflow and activity inner APIs, local activities, async activity completion, etc
+- Phase 3 — payload converters, error handling, middleware
+- Phase 4 — Sandbox, replayer, support for testing
+
+
+## Overview
+
+The Ruby SDK will be implemented on top of the [SDK Core](https://github.com/temporalio/sdk-core).
+The exact bridging mechanism will decided later as there are different alternatives.
+
+### Goals
+
+- Quickest path to a working setup
+- Convention over configuration
+- Avoid sacrificing configurability of a complex setup
+- Minimise external dependencies (gems)
+- Local and end-to-end testability of workflows
+
+### Non-goals
+
+- Replicate the interface of other SDKs
+- Backwards compatibility with Coinbase's Ruby SDK
+
+### Tech stack && approach
+
+- Ruby >= 2.7 (currently oldest fully supported version)
+- Packaged as a Ruby gem (and published to [https://rubygems.org])
+- FFI or Rutie bridge into SDK Core (to be decided)
+
+
+## An end-to-end example
+
+Perhaps a bit unconventionally (as far as the proposal review process goes), but we'd like to start
+with an end-to-end example of a simple and fully functional "Hello World" workflow. The intention of
+this example is to demonstrate how everything ties together before diving into each separate moving
+part and exploring it in depth.
+
+```ruby
+require 'temporal'
+require 'temporal/worker'
+
+Temporal.configure do |config|
```

Comment:
I understand the negative implications of a global state, however this is quite a common pattern in the Ruby ecosystem ([Sidekiq](https://github.com/mperham/sidekiq/wiki/Using-Redis#using-an-initializer), [Rails](https://guides.rubyonrails.org/configuring.html#configuring-middleware), [Bugsnag](https://docs.bugsnag.com/platforms/ruby/rails/#tracking-releases), [RSpec](https://relishapp.com/rspec/rspec-core/v/3-0/docs/configuration/custom-settings), [Kaminari](https://github.com/kaminari/kaminari#general-configuration-options), etc).

The idea here is a global mutable state, but rather a shared config. One other insight into Ruby that makes this a necessity (in many cases) is that all `require`s are global (unlike Python or Typescript). Which means that devs in their applications can't require a config class and assign it to a local variable. Once it's required, it's global.

I'm not suggesting this is a good design choice for the language, but it's a limitation we have to work with. If we disallow global `Temporal.configure`, people would end up implementing the same thing themselves, because they have to somehow share the same configuration when initializing client instances.

We can swim against the current and remove this, but I don't think people will love our SDK for that.

But as you've probably noticed further down, there is a way to do this should you need to. If you need to connect to 2 or more servers — you can do so by initializing clients yourself (in fact we did this at CB as a horizontal scaling tactic). The truth is — most application won't ever need it.

Also note that unlike other SDKs the Client here doesn't restrict your connection to a single namespace. You can use the same Client to schedule Workflows across as many namespaces as you'd like.

Reply:
If this is something common in Ruby I'd be fine with it but I'm not sure I'd consider this a core part of the proposal.